src/newcharacter.cpp

```diff
@@ -1503,8 +1531,58 @@ int set_scenario(WINDOW *w, player *u, int &points)
     ctxt.register_action("NEXT_TAB");
     ctxt.register_action("SORT");
     ctxt.register_action("HELP_KEYBINDINGS");
+    ctxt.register_action("FILTER");
+
+    bool recalc_scens = true;
+    int scens_length;
+    std::string filterstring = "";
+    std::vector<const scenario *> sorted_scens;
 
     do {
+        if (recalc_scens) {
+            sorted_scens.clear();
+            for (scenmap::const_iterator iter = scenario::begin(); iter != scenario::end(); ++iter) {
+                if (!filterstring.empty()) {
+                    if (!lcmatch(iter->second.gender_appropriate_name(u->male), filterstring)) {
+                        continue;
+                    }
+                }
+                sorted_scens.push_back(&(iter->second));
+            }
+            scens_length = sorted_scens.size();
+            if (scens_length == 0) {
+                popup(_("Nothing found.")); // another case of black box in tiles
+                filterstring.clear();
```

Comment:
> Now this brings a question of whether it would be a good idea to sync all filters to the same UI - methinks that yes :)

Yes, and this "same UI" you propose already exists, it's the ... `uimenu`. Seriously: the menu already supports filtering entries, it supports hotkeys, differently colored text, callbacks, draws scroll bars, manages the currently selected index (keeps it through filtering), has nice key bindings (page up/down), etc.

It has all the stuff that is currently repeated all over. Or that is missing in the various lists.

The wish item, spawn monster and toggle mutations debug things are implemented using the `uimenu`.

One would probably need some adjustments for special filters (the menu only filters based on the displayed text). And sorting is not supported. And you might need to add code to prevent the border.